### DISCUSSION TITLE
Adding support for EDI (X12 and EDIFACT) alongside JSON, XML, ...

### DISCUSSION THREAD

**Discussion by canabrook** [atlasmap/atlasmap#4398] [Ideas]
I have created a proof-of-concept/prototype of an EDI module analogous to the JSON, XML, ... modules and am pleased with the result. It uses the EDIReader Framework that underlies https://github.com/BerryWorksSoftware/edi-json, but supports direct creation of an AtlasMap Document from the EDI without first converting it into JSON or XML. This provides for an EDI-friendly appearance in the mapping UI, and allows efficient transformations between EDI, JSON, and XML. 

At the moment, my EDI module is is replacing the CSV module, allowing me to focus on the Java handling of the content without having to modify the UI at all. This is OK for a proof-of-concept, but obviously not beyond that.  (I am a Java developer, but not proficient with the React-based coding of the UI.)

I am looking for ideas/suggestions on how to turn this prototype into an actual thing. I could create a fork, but I would prefer something cleaner. It would be nice if the standard UI had a way to configure an optional module that might come from a 3rd party, or even documentation on exactly what needs to change in the UI to add another module. I would be willing to release an open source version of the EDI service that allows some truly useful EDI mapping/translation features, and contribute that to the atlasmap project, and also have a "premium edition" of the EDI service that can be added to the classpath for easy inclusion in any AtlasMap environment. Perhaps there are better options that I have not considered.

If anyone knows of other efforts to integrate EDI as a first-class data format alongside XML, JSON, CSV, I would be interested. No need for me to do this if it already exists. I have worked with users of my EDI-JSON tools who want the JSON so that it can be fed to AtlasMap; things would be a lot better for them if AtlasMap could take the EDI natively. And the modular design of AtlasMap already allows this ... almost.

**Comment by martywaz** [comment]
Were you ever able to make headway with this?  We're thinking about enhancing our mapper with EDI as well

**Comment by canabrook** [comment]
Hi,
I got busy with other projects, and put this one aside. Not because I hit a dead-end or decided it was a bad idea, but primarily because I needed to modify the "front-end" UI as well as the “back-end” Java. I would be willing to share notes on how I was approaching the Java side of things, making use of the EDIReader EDI parser that I created some years ago. 

Actually, I may have a fresh project coming up requiring mapping/transformation to/from EDI and JSON. I know I can use my EDI framework to convert the EDI to either JSON or XML, and then use off-the-shelf AtlasMap techniques to transform to the custom JSON needed by the application. But it would be nice to avoid the double conversion.

Scott

**Comment by canabrook** [comment]
I would be very interested in discussing options for collaborating on this effort. Please contact me by email ***@***.*** ***@***.***> or ***@***.*** ***@***.***> 

Scott
